Send the mime icon if we can't generate a preview

oparoz opened the pull request:
Sending an invalid OC_Image instance can lead to all sorts of unwanted side effects with 3rd party apps.
The first step to fix this would be to send the mime icon.
Later on we could maybe generate a broken image to send back, to indicate that we were unsuccessful in generating a preview.

It's also possible for individual providers to do a better job by generating their own broken image, but I thought it would be best to have a global solution in place first.

fix #14106

oparoz commented:
I think that we should also send a 415 or something when the preview generation fails. That way any app which expects a preview can deal with the exception properly.

PVince81 commented:
Strange thing that we didn't send the mime icon before. @icewind1991 do you know why ?

Ideally the JS/client logic should not need to think about whether to load a preview or not and simply connect to the preview URL and get whatever is supposed to be displayed, which I believe this PR achieves.

DeepDiver1975 commented:
Shouldn't we load mime icons via url to allow static serving by the web server? 

With this approach we will load the mime in php to ram and dump it to the network buffer.

Sounds pointless. 

Further more the preview url is generating previews. If previews are to available -> tell the client to react.

Mime icons are no previews.

PVince81 commented:
Hmm ok, so we need to keep `isPreviewAvailable`. I was hoping to kill it somehow :smile:

oparoz commented:
@PVince81 - `isPreviewAvailable` is very useful for JS scripts added to the Files app
@DeepDiver1975 - I agree. Loading a static URL to the mime icon would be great. We need 2 things:
* A public JS function which loads the svg or png icon, based on a mimetype
* A returned error code so that the client knows that the preview generator has failed to generate one. I don't see the point in sending an empty Image object, like the class is doing now.

icewind1991 commented:
There is an optional argument that can be passed to to preview.php `forceIcon` which when set to false will return a 404 when a preview is not available and the web interface will fallback to the mime icon.
This can be extended to cover other cases where we cant generate a preview

This currently defaults to true due to backwards compatibility reasons.

schiesbn commented:
I tried it with the version drop-down where we have a similar issue, see #14106 

Your code gets execute but the icon is still broken for a odt file. Can you have a look at it? Thanks!

oparoz commented:
@schiesbn - I think it's unrelated and needs to be addressed directly in the office preview provider. Test this PR to see if I'm right: https://github.com/owncloud/core/pull/13654

schiesbn commented:
@oparoz same result with #13654, the icon is still broken. It also isn't about office documents in general but about any document type where we can't create a preview. In this case we should show the mime-icon and I think that's what you PR does, right.

Also your code in general looks good. In my case the preview is null and I execute the if statement. 
I had a more detailed look at the icon path and saw that the path is broken. It returns

"/home/schiesbn/public_html/owncloud/master/~schiesbn/owncloud/master/core/img/filetypes/x-office-document.png"

correct would be:
"/home/schiesbn/public_html/owncloud/master/core/img/filetypes/x-office-document.png"

This is the URL I use to access ownCloud "http://localhost/~schiesbn/owncloud/master"

It seems like https://github.com/owncloud/core/pull/12546/files#diff-18fbdb7eabef5315cacf23e5a9dd6970R547 combine both paths to some extend.

oparoz commented:
Thanks for doing all the leg work @schiesbn ! :)
Seems to me that `OC::$WEBROOT` is broken and could lead to other problems elsewhere.

schiesbn commented:
@oparoz I don't think that ```OC::$WEBROOT``` is broken, it does what it should. It gives you the path relative to your domain. The Problem is that what you want is to result of ```\OC_Helper::mimetypeIcon($this->mimeType);``` but with ```\OC::$SERVERROOT``` instead of ```OC::$WEBROOT```, at the moment you concatenate both.

I think you need to do one of these things:
(a) to strip replace ```OC::$WEBROOT``` with  ```\OC::$SERVERROOT``` from the result of ```\OC_Helper::mimetypeIcon($this->mimeType);``` (easy but not the cleanest solution)
(b) extend ```\OC_Helper::mimetypeIcon($this->mimeType);``` with a parameter to decide if you want the path with  ```OC::$WEBROOT``` or  ```\OC::$SERVERROOT```
(c) come up with a new method as a replacement for your ```\OC_Helper::mimetypeIcon($this->mimeType);```  call.

> or your config is not supported

I think that's a completely valid setup.

oparoz commented:
@schiesbn - Thanks for laying it down for us. I finally took the time to look at this and I agree that I did it wrong :flushed: .

I'll discard (c) as I don't think it would be wise to duplicate that method and between (a) and (b) I'll chose (a) because this seems to be the only place where this is needed.

We will be able to revert back to sending empty objects once there is a public Javascript method which returns the relative path to mime icons.

So, if I get 2 :ok: for 
> replace OC::$WEBROOT with \OC::$SERVERROOT from the result of \OC_Helper::mimetypeIcon($this->mimeType);

I'll modify this PR.

schiesbn commented:
@oparoz I think this would be OK in this case

oparoz commented:
One more :ok: to go with that from @icewind1991 or @PVince81 ?

schiesbn commented:
@oparoz I still think that's the way to go. Do you think you can update the pull request the next few days/weeks. This bug goes down to OC6 as far as I know and it would be great if we could fix it with OC8.1 which will be released at the end of May.

oparoz commented:
@schiesbn - Yes, not a problem

schiesbn commented:
@oparoz great... Looking forward to it!

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/81f19f1d-5c7f-42b0-80a2-e0a747122cf7) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/81f19f1d-5c7f-42b0-80a2-e0a747122cf7) completed: **5 new issues**

oparoz commented:
@schiesbn - Let me know if this now works for you.

schiesbn commented on lib/private/preview.php:
```diff
@@ -524,6 +524,10 @@ public function getPreview() {
 			$this->preview = new \OC_Image();
+			$mimeIconWebPath = \OC_Helper::mimetypeIcon($this->mimeType);
+			$mimeIconServerPath =  str_replace(\OC::$WEBROOT, \OC::$SERVERROOT, $mimeIconWebPath);
+			$icon = \OC::$SERVERROOT . $mimeIconServerPath;
```
now we have two times the ```\OC::$SERVERROOT``` in the path, you replace ```\OC::$WEBROOT``` with the server root on line 528 and add it again on line 529

oparoz replied:
Of course... I've pushed an updated version

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/a38c01db-f4cf-4520-bde4-27245f5f872d) was created.

schiesbn commented:
tested and code looks good :+1: 

(just a small issue: You could update the comment before the if-statement to describe what we actually do https://github.com/owncloud/core/pull/12546/files#diff-18fbdb7eabef5315cacf23e5a9dd6970L524)

schiesbn commented:
need a second reviewer... maybe @PVince81 ?

georgehrke commented:
will test

oparoz commented:
> just a small issue: You could update the comment before the if-statement to describe what we actually do 

Absolutely

georgehrke commented:
Tested and works :+1: 

One point though:
@danimo @rperezb How will the clients behave if the preview has not the expected size?

oparoz commented:
@georgehrke - Good point. I think it would only be a problem for browsers which can't display SVGs.

oparoz committed:
Send the mime icon if we can't generate a preview

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/eb21ae5d-8e6a-44bd-a192-2d8920c44b1f) was created.

oparoz commented:
Comment has been updated

oparoz commented:
@danimo @rperezb - What's the verdict? Are your apps affected if you get a mime icon instead of a preview?

davivel commented:
@georgehrke , @oparoz , need to check it; busy day, probably until afternoon won't be able to do it

schiesbn commented:
I just talked to @dragotin. It shouldn't be a problem for the desktop client.

georgehrke commented:
ping @davivel any update? :)

davivel commented:
Yes, just working on it.

The iOS app still does not have remote thumbnails, so no problem there.

In Android we get the mime icon, that is shown with an extra black background in the list view and in a horrible blurry format in the grid view. Would need to fix it.

I tried to pass the parameter forceIcon=false that @icewind1991 connected, but it's not working (I get the mime icon anyway). Maybe I sent it wrong? Just triggering a GET to : http://<domain>/apps/files/api/v1/thumbnail/384/384/Photos/broken.jpg?forceIcon=false

PVince81 commented:
So does it mean we can merge this ? Or not ?

davivel commented:
IMHO, :-1: 

This breaks the behaviour of the remote thumbnails API in http://[domain]/apps/files/api/v1/thumbnail/ .

Wee need some parameter like the described in https://github.com/owncloud/core/pull/12546#issuecomment-73516739 that can be entered in the thumbnails entry point, allowing to the clients receive an error instead of a default icon.

Another valid option for us would be returning an icon with the requested resolution.

schiesbn commented:
@georgehrke @oparoz  would it be hard to add such a option? Or maybe even better scale the mimeicon to the requested size the same way we would do it with a "real" preview, afaik they are available as svg?

georgehrke commented:
> afaik they are available as svg?

Would create a hard dependency on imagick, which is something we don't want.

schiesbn commented:
> Would create a hard dependency on imagick, which is something we don't want.

But we already depend on it for the real previews, otherwise we couldn't provide previews in the requested size. Right? If previews are enabled we need imagick so we can also use it to scale the mime icons. If previews aren't enabled we don't execute the code anyway. Do I miss something?

georgehrke commented:
> But we already depend on it for the real previews, otherwise we couldn't provide previews in the requested size. Right?

We use imagegd for that, which was a hard dependency already before we had the preview system.

DeepDiver1975 commented:
There is no need to scale mime type icons as they are SVGs.

georgehrke commented:
For the web-gui yes, but we are talking about the thumbnail api for clients, which is supposed to serve png

davivel commented:
In that case, what about the option of adding the parameter to avoid the fallback to the icon?

DeepDiver1975 commented:
> For the web-gui yes, but we are talking about the thumbnail api for clients, which is supposed to serve png

Not true - clients can render SGVs as well. Never the less: serving mime icons is not serving previews - that's why I said in the very beginning: if there is not preview -> send back error.

oparoz commented:
@davivel @DeepDiver1975  - Sending an error code is the best option and it can easily be added to the API endpoint (same code as in preview.php), but as I said before, OC doesn't have a public JS method which can be used by the WebGUI  and apps to show the file's mimeicon.

MorrisJobke commented:
@owncloud-bot retest this please

karlitschek commented:
@oparoz @DeepDiver1975  What is the status here? Can we merge?

davivel commented:
Nothing changed on our side. If this is merged, ownCloud for Android will show very blurry mime type icons for images that cannot be previewed. 

As far as I understand, this is a lateral effect due to the interaction between the preview generation and the remote thumbnails API. I would fix it myself if I thought I could do it in a reasonable time, but I know nothing about the code in core.

oparoz commented:
While working on fixing Preview, I just had some thoughts.
Imho, even if sending the mime icon is not the long term solution, it's the client which needs fixing.
If an Android tablet asks for a 2K picture of a 50x50 animated GIF, I sure hope the client is not expecting to receive a 2K picture full of empty space. That's a lot of waste.
So similarly, if the client receives an icon, it should be able to place it in the middle of some scaffolding.

What do you think @davivel?

davivel commented:
Sure, we can center the image and fix the default background color so that it matches the white background of the list. But what about the very blurry images in the grid view.

The client is asking for a picture of certain resolution. If the server is not able to return a picture with the requested resolution, or with a resolution easily embeddable in it, an error message should be returned, not a default image of a completely different resolution.

oparoz commented:
Ah yes, I was only thinking about large previews.
So currently you get a blank preview and I'm guessing that the grid view is able to discard them because they're zero sized...

oparoz commented:
Just had a thought. Once the fixed Preview class has been merged, I should be able to generate a mime icon, surrounded by empty space to match the asked dimension.
Would that be an acceptable temporary solution @davivel?

nickvergessen commented:
So whats the status here? I'd still like to see this resolved, in order to fix the broken previews in the version dropdown for .odt files before 8.1, see #14106

@DeepDiver1975 @schiesbn review time?!

:+1:

georgehrke commented:
@nickvergessen I think we still have concerns regarding the previews in mobile clients.

@davivel @oparoz Am I right?

oparoz commented:
There are 2 public API issues which needs to be sorted
* clients need to receive a proper status code
* there needs to be an endpoint serving icons (my guess). 

Regarding the WebGUI
* we need the controllers to return proper error codes when they encounter a broken preview
* we need @rullzer's PR https://github.com/owncloud/core/pull/16724 to be fixed and merged, so that the Web GUI can retrieve the icon when it gets an error code

That probably wont happen before 8.2

In the meantime, 2 things can happen:
* API clients can check what they get and act accordingly
* Preview could return an icon surrounded by empty space so that the dimension of the preview match the asked dimension. Should be dead easy to do.

DeepDiver1975 commented:
> @nickvergessen I think we still have concerns regarding the previews in mobile clients.

indeed :-1: as long as this holds a regression for the mobile clients

oparoz committed:
Move media type icon preview creation to its own method

oparoz committed:
Remove unneeded returns from private cropping methods

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/f01caa47-7e4a-4858-aa54-25c6d3a50d0a) was created.

oparoz commented:
So, in order to temporarily accommodate mobile clients, the icon is now surrounded by empty space.

I've moved the logic to a new method and added a test.

@davivel - Please confirm that you no longer get a stretched icon
@schiesbn - Please confirm that you get the icon in versions. #worksforme

oparoz commented:
Other people can review as well @PVince81 @georgehrke @DeepDiver1975 

### Web GUI

1. Make sure documents is installed
2. Upload an odt file
3. Click on it to make a modification
4. Make the modification and go back to Files
5. Click on versions

You should be seeing the document mime icon

### Mobile
No idea how to test

georgehrke commented:
recent code changes look good,
issue with broken preview in version dropdown is fixed for me:
![preview_versions](https://cloud.githubusercontent.com/assets/1250540/8230667/b87fdcfe-15c2-11e5-981f-53c54f02bf83.png)
![preview_versions_2](https://cloud.githubusercontent.com/assets/1250540/8230668/bc6d48ba-15c2-11e5-9bc5-efe341225689.png)

DeepDiver1975 commented:
Thanks a lot @oparoz - @davivel can you please test this with mobile? THX a lot!

oparoz commented:
Since the mobile app supports multiple accounts (kudos for that!), I was able to test this on Android and the one problem we're facing is that the icon is tiny on high DPI devices, because the Preview class has no way of knowing which ratio was applied.

I don't know how to fix this without touching the API.

davivel commented:
Hi, everybody.  I'll book some time tomorrow for deep review in this problem. Thanks for your patience and your attention to the mobile apps in this issue.

schiesbn commented:
tested, works great :+1:

oparoz commented:
That's what happens on mobile... ;(
![img_20150618_163150](https://cloud.githubusercontent.com/assets/323220/8243333/786a0f46-1616-11e5-81de-3545fffaf9d6.png)

So, to me it looks like we have to pick, for 8.1, if we want to sacrifice the web or the mobile thumbnails.

On 8.0, mobile works, but not the web. To which extent? I don't know. Versions are affected, what else? Activity? The bin? They all use different end points which need fixing until there is a universal endpoint or API.

MorrisJobke commented:
@oparoz Alternative would be the use of something like owncloud/core#16724 which would be the way to go....then we have the mimetype icons by default and load thumbnails if they are available

DeepDiver1975 commented:
What about fixing this in the version's preview endpoint only?

https://github.com/owncloud/core/blob/master/apps/files_versions/ajax/preview.php#L48-L56

It should be possible to check if the preview image is valid and send the mime type icons.

schiesbn commented:
@DeepDiver1975 I think it is a general discussion about how we want things to work. Yes, we could fix it for versions but then you can wait until we find the next place where we have to fix it, and the next place, and....

From my understanding a app shouldn't care about it. It should ask for a preview and should get the best possible result, which in my opinion is either a preview of the current content or if not possible at least a mime icon.

nickvergessen commented:
> @DeepDiver1975 I think it is a general discussion about how we want things to work. Yes, we could fix it for versions but then you can wait until we find the next place where we have to fix it,

On the other hand fixing one place so close to the release is okay, and then do the general solution for the next feature release?

georgehrke commented:
> On the other hand fixing one place so close to the release is okay, and then do the general solution for the next feature release?

I second that.

> A app should be able to ask for a preview and get the best possible result, which in my opinion is either a preview of the current content or if not possible at least a mime icon.

The point about not serving a mine icon was that we want to use svg for displaying mime-icons. (so it's not blurry on high-resolution displays). The preview api endpoint is only supposed to serve png, not svg.

DeepDiver1975 commented:
> which in my opinion is either a preview of the current content or if not possible at least a mime icon.

No - a mime icon is not a valid replacment for a preview. Just have a look at the discussion - e.g. the mobile apps have their own handling.



> I think it is a general discussion about how we want things to work.

Yes - but what is the result now after months of discussion and two weeks before the release: NONE

davivel commented:
I tested the update. My opinion is still :-1:

There is a problem in mobile we need to fix on our side when the recieved preview is smaller than expected (because the image is really smaller, for example). The problem is the same with server in master as with this PR. We'll take care.

But there is no reason to accept a remote icon as a thumbnail in the mobile app. That's what the app receives when the image file, for instance, is damaged. Put aside the issue of the size. If the user has a folder with hundreds of images, the server will return hundreds of times the same icon. That's not acceptable, specially for a native app that holds its own set of icons. 

Let me remind you, in case it helps, that we are not requesting directly to preview.php, but to the entry point /index.php/apps/files/api/v1/thumbnail/ . I guess it's calling to preview.php indirectly, but I don't know the details. Maybe @tobiasKaminsky can help with this part, if I'm not wrong he contributed to the code in that entry point.

oparoz commented:
I agree with @davivel. On mobile, or even in apps, you don't want to download the same mime icon over and over. The app or browser can't cache it.
This was meant as a stop-gap measure because some endpoints would return a 200 status and no preview.

The solution for 8.2 has already been discussed:
1. Send back a proper HTTP status code
2. Let the app use what's available to retrieve the appropriate icon

For 8.1, we should make getMimeIcon public (just like it exists in JS) and there are 4 files which need to be modified
- [ ] [apps/files_versions/ajax/preview.php](https://github.com/owncloud/core/blob/c700f42b68f430e9c89ce97b92ea91323c9f6ed5/apps/files_versions/ajax/preview.php)
- [ ] [core/ajax/preview.php](https://github.com/owncloud/core/blob/c700f42b68f430e9c89ce97b92ea91323c9f6ed5/core/ajax/preview.php)
- [ ] [apps/files_trashbin](https://github.com/owncloud/core/blob/c700f42b68f430e9c89ce97b92ea91323c9f6ed5/apps/files_trashbin/ajax/preview.php) 
- [ ] [apps/files_sharing/ajax/publicpreview.php](https://github.com/owncloud/core/blob/8d4f918671266b0b0457335f6fa56eb23841ae04/apps/files_sharing/ajax/publicpreview.php) to call the same method.

oparoz commented:
I've found an easier way which won't impact the public API. I'll let you know if it breaks other stuff.

oparoz committed:
Remove unneeded returns from private cropping methods

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/74be58db-8fbe-4619-9971-4e63a4348ffb) was created.

oparoz commented:
Wrong commit message... but I'll fix it in the squash.

All 4 files mentioned earlier rely on `showPreview()`, which is only used by web clients, as it includes HTTP headers, so I've modified that method to include the fallback.

- [ ] versions get their icons
- [ ] bin gets its icons. Not sure when it did not. In my tests with odt files, I could see the icons.
- [ ] single file sharing shows the 36x36 fallback icon in case something went wrong with the preview. Otherwise, it shows the 128x128 svg files
- [ ] filelist shows thumbnails
- [ ] mobile client uses its own icons since it never receives the fallback icon

oparoz commented:
In order tu support high DPI devices, we should be using SVGs, but we can't because we need to support IE8 in core.

georgehrke commented:
I would actually prefer to go with https://github.com/owncloud/core/pull/17042.
As @DeepDiver1975 said, mime type-icons ain't previews, and the preview api should only return something, if it actually generated a preview.

As this is something we need to fix with js and proper response codes (in 8.2), we should only change as little code as needed to the unwanted "return a png, no matter what"-logic.

> we should be using SVGs here, but we can't because we need to support IE8 in core.

We already have some JS-code in place, that automatically replaces png mime-type icons with svg-icons, given that the browser supports it. That's yet another reason why the preview api shouldn't return mime-type icons.

oparoz commented:
#17042 is just pushing the icon preview generation to the endpoint. It's still wrong, as we're still in PHP and in a preview endpoint. On top of that, you have to have the exact same code in 4 different locations.

My suggestion is to only touch one location before properly fixing the JS. Those endpoints don't need any attention right now.

davivel commented:
Just tested, the Android app works fine with this update. Also with #17042.

DeepDiver1975 commented:
I'm fine to move this in 8.1 - @schiesbn mind retesting versions? Thanks a lot!

@oparoz Thanks to you for digging into this issue for all that weeks! Really awesome!

schiesbn commented:
Tested with versions, works as expected :+1:

schiesbn commented:
> @oparoz Thanks to you for digging into this issue for all that weeks! Really awesome!

I agree, great job!

oparoz commented:
You're welcome!
